specs/AppLifecycle/Instancing/AppLifecycle SingleMulti-Instancing.md

```diff
@@ -0,0 +1,422 @@
+            RegisterClassAndStartMessagePump(hInstance, nCmdShow);
+        }
+    }
+    return 1;
+}
+```
+
+### Handling redirection
+
+The first 2 scenarios illustrate how an app would redirect to another running instance. This
+scenario shows how an app would handle the case where it is already running, and then a fresh
+activation is redirected to this instance. The AppInstance class will expose an Activated event, for
+which the app can register. In its Activated event handler, the app would examine the type of
+activation, retrieve the rich ActivationArguments object, and then do whatever app-specific work it
+wishes to do based on those arguments.
+
+```c++
+int APIENTRY wWinMain(
+    _In_ HINSTANCE hInstance, _In_opt_ HINSTANCE hPrevInstance,
+    _In_ LPWSTR lpCmdLine, _In_ int nCmdShow)
+{
+    // First, hook up the Activated event, to allow for this instance of the app
+    // getting reactivated as a result of multi-instance redirection.
+    AppInstance::Activated([](ActivationArguments args)
+        { OnActivated(args); });
+
+    //...etc - the rest of WinMain as normal.
+}
+
+void OnActivated(ActivationArguments const& args)
+{
+    ExtendedActivationKind kind = args.Kind();
+    if (kind == ExtendedActivationKind::Launch)
+    {
+        auto launchArgs = args.Data().as<LaunchActivatedEventArgs>();
+        DoSomethingWithLaunchArgs(launchArgs.Arguments());
+    }
+    else if (kind == ExtendedActivationKind::File)
+    {
+        auto fileArgs = args.Data().as<FileActivatedEventArgs>();
+        DoSomethingWithFileArgs(fileArgs.Files());
+    }
+    else if (kind == ExtendedActivationKind::Protocol)
+    {
+        auto protocolArgs = args.Data().as<ProtocolActivatedEventArgs>();
+        DoSomethingWithProtocolArgs(protocolArgs.Uri());
+    }
+    else if (kind == ExtendedActivationKind::StartupTask)
+    {
+        auto startupArgs = args.Data().as<StartupTaskActivatedEventArgs>();
+        DoSomethingWithStartupArgs(startupArgs.TaskId());
+    }
+}
+```
+
+### Redirection orchestration
+
+We can also imagine a more advanced scenario, where the app registers an instance for redirection,
+but when it receives a redirected activation, it may apply further app logic to decide whether to
+handle this new activation in the current instance or redirect again, effectively chaining
+redirections.
+
+```c++
+void OnActivated(ActivationArguments const& args)
+{
+    const ExtendedActivationKind kind = args.Kind();
+
+    // For example, we might want to redirect protocol activations.
+    if (kind == ExtendedActivationKind::Protocol)
+    {
+        auto protocolArgs = args.Data().as<ProtocolActivatedEventArgs>();
+        Uri uri = protocolArgs.Uri();
+
+        // We'll try to find the instance that handles protocol activations.
+        // If there isn't one, then this instance will take over that duty.
+        auto instance = AppInstance::FindOrRegisterForKey(uri.AbsoluteUri().c_str());
```

Comment:
```suggestion
        auto instance = AppInstance::FindOrRegisterForKey(uri.AbsoluteUri());
```

No need to convert from `hstring` to `wchar_t const*` then back to `hstring`.

Reply:
RECOMMEND: Fix

Reply:
Fixed.